### ISSUE TITLE
getCurrentStatus() not working

### ISSUE THREAD

**Issue by zmip** [firstred/postnl-api-php#18]
I've spent a few hours deciphering what happens when calling getCurrentStatus() with a barcode. The data returned from the API is valid, but things seem to go horribly wrong at parsing the JSON data into the appropriate data structures. Docs are unclear at what should happen, but I assume the returned object from getCurrentStatus() (a CurrentStatusResponse) should be hydrated with data. It's not. There's only one fillable prop on CurrentStatusResponse ($Shipments) and that doesn't get touched. Makes sense, in the returned JSON there is no "Shipments" key, only a "Shipment" key, but this gets entirely ignored AbstractEntity's jsonDeserialize(). I assume that there should actually be a property $CurrentStatus on CurrentStatusResponse (so you can call getCurrentStatus() on it). If I add that, then it gets filled, but all the other data from the API call is not getting attached to it. 

So what happened here? Did PostNL change the returned JSON, or did this never work, and am I the only actually trying to get this working?

**Comment by firstred** [comment]
> So what happened here? Did PostNL change the returned JSON, or did this never work, and am I the only actually trying to get this working?

Yes, the REST API has changed so often that now the original v1 version is broken for the most part. The SOAP version shouldn't be used for new projects, as PostNL is phasing that one out.

We have been trying to keep up with a new version v2, but eventually had to abandoned that one, too. The API is changing rapidly without PostNL publishing what exactly has changed. Many API components are an outright mystery as the documentation is either lacking or outdated. For now my (and a few others') API "fix" has been to switch to DHL, because they take their API far more seriously. I am hoping to reboot this project this year, because PostNL still is a major player and I can't believe the API will stay broken for another year.

For the moment I would strongly discourage people from making their business depend on this unstable API. Try to find an intermediate shipping tool such as Sendcloud, Wuunder, etc. so you wouldn't have to be in the same sh*t I have been with PostNL.

**Comment by zmip** [comment]
Wow, thanks for sharing your insights Michael. Pretty amazing actually. Isn't that exactly why API's have a version prefix? To **not** break other people's applications when changes are made?

I've never seen such amateuristic and incomplete API documentation as at PostNL's. Actually, make that: never seen such clumsiness as an entire business as PostNL. Rather pathetic party to work with, their IT-related stuff is a total laugh. But on the plus side, the people are super friendly (I had to find out that this is the type of IT (!) company that you have to actually phone to get stuff done, haha! And then they covertly smirk at the incompetence of their own company...).

So even more kudos to you for actually having created something that works/worked. I would have probably given up before even starting, just looking at the horror that is their documentation. I've implemented a quick workaround by subclassing the main ThirtyBees\PostNL class, but I'll probably roll something myself or, like you say, move to DHL or DPD.

**Comment by firstred** [comment]
Before you are going to spend more time on it, let's see if I can find a dump from a store that still ships with PostNL. I believe there's one that has a version that has been hotpatched over the years. Absolute madlad who kept supporting this!
It has been scoped with [PHP-scoper](https://github.com/humbug/php-scoper) to not collide with other PostNL modules in a sloppy e-commerce system; the base namespace has become `ThirtyBeesPostNL\ThirtyBees\PostNL`, but other than that it is still compatible with this documentation: https://postnl-php.readthedocs.io/en/1.1.0/
It has been running in production now for a whopping 3 months without any breaking changes.

**Comment by firstred** [comment]
Here it is:
[postnlvendor.zip](https://github.com/firstred/postnl-api-php/files/4082879/postnlvendor.zip)

You will have to add it to your composer file manually:
```
{
    ...
    "autoload": {
        "classmap": [
            "path/to/thirtybees/postnl-api-php"
       ],
    ...
}
```
but then you should be able to use it right away.
The developer settled with the SOAP API because that one has been the most trustworthy of the two. It is also likely the only version that is supported.

Another option is to use the SOAPClient helpers from this library: https://github.com/DivideBV/Postnl, but note that it doesn't support concurrent connections (might be slow at times). It also doesn't come with the abstractions this library provides and has some more low-level annoyances, but since it's closer to the API, it may be a lot easier to detect changes.

I hope this has been helpful in some way. Good luck with your project(s)!

**Comment by firstred** [comment]
Should have been fixed by 1.2.0
